src/dbnode/storage/bootstrap/bootstrapper/uninitialized/source.go

```diff
@@ -0,0 +1,172 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package uninitialized
+
+import (
+	"github.com/m3db/m3/src/dbnode/storage/bootstrap"
+	"github.com/m3db/m3/src/dbnode/storage/bootstrap/result"
+	"github.com/m3db/m3/src/dbnode/storage/namespace"
+	"github.com/m3db/m3/src/dbnode/topology"
+	"github.com/m3db/m3cluster/shard"
+)
+
+// The purpose of the unitializedSource is to succeed bootstraps for any
+// shard/time-ranges if the given shard/namespace combination has never
+// been completely initialized (is a new namespace). This is required for
+// allowing us to configure the bootstrappers such that the commitlog
+// bootstrapper can precede the peers bootstrapper and still succeed bootstraps
+// for brand new namespaces without permitting unintentional data loss by
+// putting the noop-all or noop-none bootstrappers at the end of the process.
+// Behavior is best understood by reading the test cases for the test:
+// TestUnitializedSourceAvailableDataAndAvailableIndex
+type uninitializedSource struct {
+	opts Options
+}
+
+// NewUninitializedSource creates a new uninitialized source.
+func newUninitializedSource(opts Options) bootstrap.Source {
+	return &uninitializedSource{
+		opts: opts,
+	}
+}
+
+func (s *uninitializedSource) Can(strategy bootstrap.Strategy) bool {
+	switch strategy {
+	case bootstrap.BootstrapSequential:
+		return true
+	}
+
+	return false
+}
+
+func (s *uninitializedSource) AvailableData(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	return s.availability(ns, shardsTimeRanges, runOpts)
+}
+
+func (s *uninitializedSource) AvailableIndex(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	return s.availability(ns, shardsTimeRanges, runOpts)
+}
+
+func (s *uninitializedSource) availability(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	var (
+		topoState                = runOpts.InitialTopologyState()
+		availableShardTimeRanges = result.ShardTimeRanges{}
+	)
+
+	for shardIDUint := range shardsTimeRanges {
+		shardID := topology.ShardID(shardIDUint)
+		hostShardStates, ok := topoState.ShardStates[shardID]
+		if !ok {
+			// This shard was not part of the topology when the bootstrapping
+			// process began.
+			continue
+		}
+
+		// The basic idea for the algorithm is that on a shard-by-shard basis we
+		// need to determine if the namespace is "new" in the sense that it has
```

Comment:
nit: instead of `namespace is "new"` --> `cluster is "new"`

Reply:
Thanks, not sure why I kept making that mistake. I guess its easy to think of the topology as being per-namespace but actually shards are for the whole cluster not a namespace